syntax = "proto2";


message Book {
  optional string title = 2;
  optional string author = 3;
  optional int32 publishedYear = 4;
  required string isbn = 1;
  optional Genre genre = 5 [default = UNKNOWN];
}

  enum Genre {
    MYSTERY = 0;
    ROMANCE = 1;
    FICTION = 2;
    AUTOBIOGRAPHY = 3;
    UNKNOWN = 4;
  }

  message InventoryItem{
    required int32 invNumber = 1;
    oneof inventoryType {
      Book book = 2;
    }
    optional Status status = 3 [default = AVAILABLE];
    enum Status {
      AVAILABLE = 0;
      TAKEN = 1;
    }
  }

  message GetBookRequest {
    required string isbn = 1;
  }

  message GetBookReply {
    optional Book book = 1;
    required string message = 2;
  }

  message CreateBookRequest {
  optional string title = 2;
  optional string author = 3;
  optional int32 publishedYear = 4;
  required string isbn = 1;
  optional Genre genre = 5 [default = UNKNOWN];
  }

  message CreateBookReply {
    required string message = 1;
  }

  service InventoryService {
  rpc GetBook (GetBookRequest) returns (GetBookReply) {}
  rpc CreateBook (CreateBookRequest) returns (CreateBookReply) {}
  }

